src/main/java/org/jenkinsci/main/modules/sshd/PublicKeyAuthenticatorImpl.java

```diff
@@ -20,27 +24,40 @@
     private final PublicKeySignatureWriter signatureWriter = new PublicKeySignatureWriter();
 
     public boolean authenticate(String username, PublicKey key, ServerSession session) {
+        boolean authenticated = this.processAuthenticate(username, key);
+
+        if(authenticated){
+            UserDetails userDetails = Jenkins.getInstance().getSecurityRealm().loadUserByUsername(username);
```

Comment:
🐛 this method may throw Runtime exceptions. I'd guess they need to be handled somehow (e.g. system log message + reject authentication).

Reply:
It also may blow up cases with `-Dhudson.model.User.allowNonExistentUserToLogin=true` and locally created users. I am not sure how to gracefully handle this case though, but it should allow such users to login even if `UserDetails` cannot be retrieved. CC @daniel-beck

Reply:
On the one hand in case of Runtime Exception, no problem to log message but the rejection would change the behavior and potentially break compatibility.

On the other hand if we want to let user without UserDetails to authenticate, I must create a new UserDetails from void for them ? or simply not triggering the event, that could be another problem in case someone relies on those calls to make modifications, audit, etc.

Reply:
You are doing too much work here, and failing to honor the existing system property. It would suffice to add somewhere in `processAuthentication`:

```java
try {
    u.impersonate();
} catch (UsernameNotFoundException x) {
    LOGGER.log(Level.FINE, username + " is not a real user", x);
    return false;
}
```

Changing the behavior is a good thing—the existing behavior is a bug.